packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -0,0 +1,189 @@
+    }
+
+    public get isEmpty(): boolean {
+        return this.attachFlowBatch.length === 0 && this.mainBatch.length === 0;
+    }
+
+    public submit(message: BatchMessage) {
+        if (!this.mainBatch.push(message)) {
+            throw new GenericError(
+                "BatchTooLarge",
+                /* error */ undefined,
+                {
+                    opSize: (message.contents?.length) ?? 0,
+                    count: this.mainBatch.length,
+                    limit: this.mainBatch.options.hardLimit,
+                });
+        }
+    }
+
+    public submitAttach(message: BatchMessage) {
+        if (!this.attachFlowBatch.push(message)) {
+            // BatchManager has two limits - soft limit & hard limit. Soft limit is only engaged
+            // when queue is not empty.
+            // Flush queue & retry. Failure on retry would mean - single message is bigger than hard limit
+            this.flushInternal(this.attachFlowBatch.popBatch());
+            if (!this.attachFlowBatch.push(message)) {
+                throw new GenericError(
+                    "BatchTooLarge",
+                    /* error */ undefined,
+                    {
+                        opSize: (message.contents?.length) ?? 0,
+                        count: this.attachFlowBatch.length,
+                        limit: this.attachFlowBatch.options.hardLimit,
+                    });
+            }
+        }
+    }
+
+    public flush() {
+        this.flushInternal(this.attachFlowBatch.popBatch());
+        this.flushInternal(this.mainBatch.popBatch());
+    }
+
+    private flushInternal(rawBatch: IBatch) {
+        const processedBatch = this.maybeCompressBatch(rawBatch);
+        const clientSequenceNumber = this.sendBatch(processedBatch);
+
+        this.persistBatch(clientSequenceNumber, rawBatch.content);
+    }
+
+    private maybeCompressBatch(batch: IBatch): IBatch {
+        if (batch.content.length === 0
+            || this.params.config.compressionOptions === undefined
+            || this.params.config.compressionOptions.minimumBatchSizeInBytes >= batch.contentSizeInBytes) {
+            // Nothing to do if the batch is empty or if compression is disabled or if we don't need to compress
+            return batch;
+        }
+
+        const compressedBatch = this.params.compressor.compressBatch(batch);
+        if (compressedBatch.contentSizeInBytes < this.params.config.maxBatchSizeInBytes) {
+            // If we don't reach the maximum supported size of a batch, it safe to be sent as is
+            return compressedBatch;
+        }
+
+        throw new GenericError(
+            "BatchTooLarge",
+                /* error */ undefined,
+            {
+                opSize: batch.contentSizeInBytes,
+                count: batch.content.length,
+                limit: this.params.config.maxBatchSizeInBytes,
+                compressed: true,
+            });
+    }
+
+    private sendBatch(batch: IBatch): number {
```

Comment:
please add comments on what this function returns (-1 or first localSequenceNumber)